File: solutions/python/3-sum.py
# 3-sum.cpp 848813e190b1b85a8e75107fe8513c3be38ad1a9
import sys
import random


# @include
def has_three_sum(A, t):
    A.sort()

    for a in A:
        # Finds if the sum of two numbers in A equals to t - a.
        if has_two_sum(A, t - a):
            return True
    return False

def has_two_sum(A, t):
    j = 0
    k = len(A) - 1

    while j <= k:
        if A[j] + A[k] == t:
            return True
        elif A[j] + A[k] < t:
            j += 1
        else:  # A[j] + A[k] > t.
            k -= 1
    return False
# @exclude


# n^3 solution.
def check_ans(A, t):
    for i in A:
        for j in A:
            for k in A:
                if i + j + k == t:
                    return True
    return False


def main():
    for _ in range(1000):
        if len(sys.argv) == 2:
            n = int(sys.argv[1])
        else:
            n = random.randint(1, 10000)
        T = random.randrange(n)

        A = [random.randint(-100000, 100000) for i in range(n)]

        print(has_three_sum(A, T))
        assert check_ans(A, T) == has_three_sum(A, T)


if __name__ == '__main__':
    main()

File: solutions/cpp/3-sum.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <random>
#include <vector>

#include "./2_sum.h"

using TwoSum::HasTwoSum;
using std::boolalpha;
using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

// @include
bool HasThreeSum(vector<int> A, int t) {
  sort(A.begin(), A.end());

  for (int a : A) {
    // Finds if the sum of two numbers in A equals to t - a.
    if (HasTwoSum(A, t - a)) {
      return true;
    }
  }
  return false;
}
// @exclude

// n^3 solution.
bool CheckAns(const vector<int>& A, int t) {
  for (int i = 0; i < A.size(); ++i) {
    for (int j = 0; j < A.size(); ++j) {
      for (int k = 0; k < A.size(); ++k) {
        if (A[i] + A[j] + A[k] == t) {
          return true;
        }
      }
    }
  }
  return false;
}

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  for (int times = 0; times < 1000; ++times) {
    int n, T;
    if (argc == 2) {
      n = atoi(argv[1]);
      uniform_int_distribution<int> dis(0, n - 1);
      T = dis(gen);
    } else {
      uniform_int_distribution<int> dis(1, 10000);
      n = dis(gen);
      uniform_int_distribution<int> T_dis(0, n - 1);
      T = T_dis(gen);
    }
    vector<int> A;
    for (size_t i = 0; i < n; ++i) {
      uniform_int_distribution<int> dis(-100000, 100000);
      A.emplace_back(dis(gen));
    }
    cout << boolalpha << HasThreeSum(A, T) << endl;
    assert(CheckAns(A, T) == HasThreeSum(A, T));
  }
  return 0;
}
